[DS-2199] DSpace stores DOIs in different metadata values depending on which DOI registration agency is used

mwoodiupui opened the pull request:
Fixes #5565 / https://jira.duraspace.org/browse/DS-2199
Including a curation task for correcting the metadata of objects that have already received DOIs through the EZID provider.

mwoodiupui committed:
[DS-2199] Use correct metadata field for generated DOI.

mwoodiupui committed:
[DS-2199] Pull duplicated metadata constants up to common superclass.

mwoodiupui committed:
[DS-2199] Document in code comments the proper metadata field for storing generated identifiers.

pnbecker commented:
The DOI is entered in the dc.identifier.uri element now, but it has the form "doi:10....". While this technically is a URI, this is not a URI most systems will be able to use as they are. Wouldn't it be better to write the URI in the form http(s)://dx.doi.org/10.... as it can be used in this form almost anywhere and is still recognized as beeing a DOI?

mwoodiupui commented:
If that is the way that the DataCite provider stores DOIs then this further change would make the two providers more consistent.  It is a reasonable suggestion, which I will implement.

We have a problem outside of the code in this area:  we have no guidance on *how an identity provider should function as a DSpace component*.  We now realize that we need agreement on where to store generated identifiers, and that we need agreement on the form in which they are stored.  I wonder what other agreements have not yet been made?

mwoodiupui committed:
[DS-2199] Store DOIs as HTTP URLs, not doi: URIs.

Replace custom formatters with DOI static methods.
Request methods now take doi: URIs, not naked suffixes.

pnbecker commented:
@mwoodiupui I agree with you, that guides or rules to harmonize this would be a good idea.

In this special case the other possibility would be to take (again?) a look on the DataCite Implementation. I tried to wrote it in a way that should make it possible to add support for the EZID API to it (see org.dspace.identifier.doi.DOIConnector). Then at least both DOI implementation would do the same for sure and EZID would use the asynchronous registration as well.

mwoodiupui committed:
[DS-2199] Return to use of supports() lost in Services API work

tdonohue commented:
@mwoodiupui : this looks to need another rebase

mwoodiupui commented:
Rebased

tdonohue commented on dspace-api/src/main/java/org/dspace/identifier/ezid/MetadataFixupTask.java:
```diff
@@ -0,0 +1,144 @@
+ */
+@Mutative
+public class MetadataFixupTask
+        extends AbstractCurationTask
```
As noted in IRC, it might be worth considering reworking this into a Flyway (Java or SQL) migration.  A curation task would "work", but a Flyway migration is more automated.  Plus, it seems like a "run once" sort of thing, which may fit better under a Flyway migration instead of a curation task.

mwoodiupui committed:
[DS-2199] Replace curation task with migration

mwoodiupui commented:
The current form of the new migration breaks many tests, apparently because the registries are not loaded at migration time when in the test environment.

mwoodiupui committed:
[DS-2199] Skip migration if the fields don't exist yet.

During testing or fresh_install, there will be no metadata fields
defined and obviously no metadata to convert.  Not bypassing this case
causes test suites to fail right and left.

mwoodiupui commented:
Surprise!  Testing for an empty MetadataSchemaRegistry table to detect fresh_install or test environment fails, apparently because one of the previous migrations registers the _dcterms_ schema.

mwoodiupui commented:
WIP because new migration is not yet tested.

mwoodiupui committed:
[DS-2199] ResultSet.first() fails on non-scrollable ResultSet.

.next() on a fresh ResultSet positions as we want and returns what we
want to know, and works on any ResultSet.

mwoodiupui committed:
[DS-2199] Add missing default row ID.

mwoodiupui committed:
[DS-2199] Commit progress on testing and fixing

mwoodiupui committed:
[DS-2199] Fix duplicate migration names.

mwoodiupui committed:
[DS-2199] Finish debugging tests, fix what they found.

mwoodiupui commented:
Rebased again.

mwoodiupui committed:
[DS-2199] Use same DataSource and Connection for the test and Flyway.

Poking at Travis to find out why it crashes here while the same code
works locally.

tdonohue commented on dspace-api/src/test/java/org/dspace/storage/rdbms/migration/V6_0_2016_04_01_0002__DS_2199_Move_IdentifiersIT.java:
```diff
@@ -0,0 +1,252 @@
+        cnctn = ds.getConnection();
+
+        // Define the database.
+        DatabaseUtilsHelpers.updateDatabase(ds, cnctn);
```
You may be able to do away with the `DatabaseUtilsHelpers` class altogether.  You can simply call `DatabaseUtils.updateDatabase()` (no arguments)

tdonohue commented:
Please rebase this PR against the latest `master` branch.  The `master` branch now requires adherence to our new Java Code Style documented at: https://wiki.duraspace.org/display/DSPACE/Code+Style+Guide

If you have any questions, feel free to ask in this PR or on our Slack: https://wiki.duraspace.org/display/DSPACE/Slack

mwoodiupui committed:
Switch SLF4J to Log4Jv2 to address ambiguous arguments. DS-2199 #1006

mwoodiupui commented:
Rebased to address conflict.

mwoodiupui committed:
Make it compile against Flyway v6.  Satisfy Checkstyle. DS-2199

mwoodiupui committed:
[DS-2199] Remove unnecessary helper class:  delegated method is now visible.

tdonohue commented:
@mwoodiupui : Revisiting this PR after a question about DOI metadata fields in today's DSpace 7.6 Q&A session.  I realized this PR seems to move all DOIs to `dc.identifier.uri`.... and we have a separate ticket #7055 which talks about moving them all to `dc.identifier.doi`.  We may need to reanalyze the best field to store DOIs in, or make it configurable.

This particular PR would also need to be updated as the Flyway migration scripts in this PR are outdated/old (and should reference DSpace 7 instead of 6).

Just wanted to note these issues in case you find time to update this for 8.0.

mwoodiupui commented:
See comments on the referenced issue.  I just thought of a refinement of my remarks there:  I propose that DSpace-generated identifiers be *stored* consistently across all DSpace instances, but that the selection of identifier types *presented* by the UI be configurable.

This patch was meant to disturb as few things as possible, but a more thorough reconsideration of the design would be welcome.

alanorth commented:
For what it's worth, regarding the format for the DOI, the "DOI Proxy" section of the [DOI handbook](https://www.doi.org/doi-handbook/HTML/index.html) says `doi.org` should be preferred to `dx.doi.org`:

> With the HTTPS Proxy Server of the DOI System (https://doi.org/), users can resolve DOI names from any standard web browser by using the URL syntax. For example, the resolution of the DOI name "10.10.123/456" would be done from the address https://doi.org/10.123/456.
> 
> NOTE  An earlier proxy address ([https://dx.doi.org](http://dx.doi.org/)) is maintained but its use is discouraged.

[Crossref has recommended](https://www.crossref.org/display-guidelines/#why-not-use-dx-as-in-httpdxdoiorg-a-id00553-href00553a) the same for years as well:

> Originally the dx separated the DOI resolver from the International DOI Foundation (IDF) website but this changed a few years ago and the IDF recommends http://doi.org as the preferred form for the domain name in DOI URLs.

tdonohue commented:
Per discussion in today's Developer Meeting, and the consensus documented in #9901, this PR will likely need to be either updated or closed (in favor of a new PR).  Any new fixes should fully align with the approach now described in #9901

bram-atmire commented:
Closing — this PR has been superseded by the DOI consensus reached in the [2024-10-17 Developer Meeting](https://wiki.lyrasis.org/display/DSPACE/2024-10-17+DSpace+Developers+Meeting), documented in #9901 and implemented by #11545 ("Unify treatment of DOIs"), which was merged on 2026-03-26.

The new approach differs from what this PR proposed:
- DataCite + EZID DOIs are now consistently stored in `dc.identifier.doi`
- Externally-generated DOIs go in `dc.relation.hasversion`
- #11545 also ships a migration tool to move existing DOIs into the new regime

Per @tdonohue's [earlier comment on this PR](https://github.com/DSpace/DSpace/pull/1006), closing this as it predates the consensus. Thanks for the original work, @mwoodiupui — much of it informed the later discussion.
